src/card_operations/card_fetch_share.c

```diff
@@ -105,16 +116,66 @@ static uint8_t remaining_cards;
 /*****************************************************************************
  * STATIC FUNCTIONS
  *****************************************************************************/
-static void _handle_retrieve_wallet_success(uint8_t xcor) {
-  if (WALLET_IS_ARBITRARY_DATA(wallet.wallet_info))
+
+static bool verify_fetched_wallet(uint8_t xcor) {
+  Flash_Wallet *flash_wallet = NULL;
+  bool status =
+      get_flash_wallet_by_name((const char *)wallet.wallet_name, &flash_wallet);
+
+  ASSERT(SUCCESS == status && NULL != flash_wallet);
+
+  bool compare_status = compare_status =
+      (0 == memcmp(wallet.wallet_id, flash_wallet->wallet_id, WALLET_ID_SIZE));
+  compare_status &=
+      (0 == memcmp(wallet.wallet_name, flash_wallet->wallet_name, NAME_SIZE));
+  compare_status &= (wallet.wallet_info == flash_wallet->wallet_info);
+
+  /**
+   * For wallets with device share present on flash, fetched wallet nonce is
+   * compared with wallet nonce on flash. In case of sync wallet, the wallet
+   * share and encryption data is written on flash after verification. For
+   * verifying the wallets in this case, we compare wallet nonce fetched from
+   * the two cards.
+   */
+  if (VALID_WALLET_WITHOUT_DEVICE_SHARE != flash_wallet->state) {
+    // Wallet nonce present on flash, so compare current wallet nonce with flash
+    // wallet nonce.
+    uint8_t wallet_nonce[NONCE_SIZE] = {0};
+    ASSERT(SUCCESS ==
+           get_flash_wallet_nonce_by_name(
+               (const char *)flash_wallet->wallet_name, wallet_nonce));
+    compare_status &=
+        (0 == memcmp(wallet.wallet_share_with_mac_and_nonce + BLOCK_SIZE,
+                     wallet_nonce,
+                     NONCE_SIZE));
+  } else if (0 < xcor) {
```

Comment:
generalize the `else-if` to `else` only.

Add assertion/separate-check on xcor instead.

Reply:
Consider the confusion when renaming as card number.
This is not technically the card number we're comparing here, of the 2 cards tapped on device, this variable represents the 2nd card tapped regardless of the #cardNumber.

Reply:
updated the doxygen comments here for the variable.